### PR TITLE
Configurable Session Token Duration

### PR DESCRIPTION
**Pull request by nathandines**
Based off of https://github.com/jenkinsci/aws-credentials-plugin/pull/43, but attempts to address the changes requested by @andresrc at https://github.com/jenkinsci/aws-credentials-plugin/pull/43#pullrequestreview-198318491

### CODE REVIEW

**Comment by nathandines** [comment]
Ping @andresrc 

Any feedback on this? Hoping to get it through soon if we can

File: src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java [hunk 1]
```diff
@@ -40,7 +40,8 @@
 import com.amazonaws.services.securitytoken.model.AssumeRoleResult;
 import com.cloudbees.plugins.credentials.CredentialsDescriptor;
 import com.cloudbees.plugins.credentials.CredentialsScope;
-import edu.umd.cs.findbugs.annotations.CheckForNull;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
```
Comment (andresrc) [inline src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: The `findbugs` annotation (even if deprecated, will be fixed by upgrading the parent pom and using `spotbugs`) are preferred for license issues.

File: src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java [hunk 2]
```diff
@@ -62,16 +64,18 @@
...
     private final Secret secretKey;
 
     private final String iamRoleArn;
     private final String iamMfaSerialNumber;
 
-    /**
-     * Old data bound constructor. It is maintained to keep binary compatibility with clients that were using it directly.
-     */
+    @CheckForNull
+    private Integer stsTokenDuration;
```
Comment (andresrc) [inline src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: This field should be protected somehow for concurrent access, as it is not final. Either make it `volatile` or guard it with a lock.

File: src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java [hunk 3]
```diff
@@ -119,7 +135,8 @@ public AWSCredentials getCredentials() {
                 initialCredentials = (new InstanceProfileCredentialsProvider()).getCredentials();
             }
 
-            AssumeRoleRequest assumeRequest = createAssumeRoleRequest(iamRoleArn);
+            AssumeRoleRequest assumeRequest = createAssumeRoleRequest(iamRoleArn)
+                    .withDurationSeconds(stsTokenDuration);
```
Comment (andresrc) [inline src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: The field can be `null`. Does this method allow a `null` parameter?

File: src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java [hunk 4]
```diff
@@ -135,7 +152,8 @@ public AWSCredentials getCredentials(String mfaToken) {
 
         AssumeRoleRequest assumeRequest = createAssumeRoleRequest(iamRoleArn)
                 .withSerialNumber(iamMfaSerialNumber)
-                .withTokenCode(mfaToken);
+                .withTokenCode(mfaToken)
+                .withDurationSeconds(stsTokenDuration);
```
Comment (andresrc) [inline src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: Ditto

File: src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java (hunk 3 above)
Comment (nathandines) [reply src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: I had thought that `defaultStsTokenDuration()` would ensure this is always populated with an integer. I might have been mistaken.
Would you recommend calling `getStsTokenDuration()` here instead?

**Comment by nathandines** [comment]
@andresrc src I've made some changes, and rebased against master to bring the parent pom forward (I didn't realise the branch was so far behind!) which I think might be better.

I did keep getting `cannot find symbol` when trying to use `edu.umd.cs.findbugs.annotations.Nonnull`. Is it a problem to use `javax.annotation.Nonnull`, or were you just referring to `CheckForNull`?

File: src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java [hunk 5]
```diff
@@ -106,6 +109,16 @@ public String getIamMfaSerialNumber() {
...
     }
 
+    @NonNull
+    public Integer getStsTokenDuration() {
+        return stsTokenDuration == null ? DescriptorImpl.defaultStsTokenDuration : stsTokenDuration;
+    }
+
+    @DataBoundSetter
+    public void setStsTokenDuration(@NonNull Integer stsTokenDuration) {
+        this.stsTokenDuration = stsTokenDuration.equals(DescriptorImpl.defaultStsTokenDuration) ? null : stsTokenDuration;
```
Comment (andresrc) [inline src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: With this logic you can relax the `@NonNull` condition, which is only used by static analysis and cannot be enforced by Jelly and write:

```
this.stsTokenDuration = stsTokenDuration == null || stsTokenDuration.equals(DescriptorImpl.defaultStsTokenDuration) ? null : stsTokenDuration;
```

File: src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java [hunk 6]
```diff
@@ -176,10 +190,13 @@ public String getDisplayName() {
             return Messages.AWSCredentialsImpl_DisplayName();
         }
 
+        public static final Integer defaultStsTokenDuration = STS_CREDENTIALS_DURATION_SECONDS;
```
Comment (andresrc) [inline src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: Do we need another constant here? If it is to reference it from Jelly, `${it.STS_CREDENTIALS_DURATION_SECONDS}` should do the trick. Anyway by convention, constants should be all uppercase.

File: src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java (hunk 5 above)
Comment (nathandines) [reply src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: Thanks for the feedback. I've added the improvements for how the method handles null values now

File: src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java (hunk 6 above)
Comment (nathandines) [reply src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: I kept the original constant for backwards compatibility, as it was public. I tried incorporating your suggestion, but could not access `STS_CREDENTIALS_DURATION_SECONDS` from Jelly using the `it` syntax, so at the moment, I've kept it as it was, but changed the casing. Please let me know if this would be okay.

**Comment by nathandines** [comment]
@andresrc I've made some changes. Awaiting feedback

File: src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java [hunk 7]
```diff
@@ -63,17 +65,18 @@
 
     private static final Logger LOGGER = Logger.getLogger(BaseAmazonWebServicesCredentials.class.getName());
 
-    public static final int STS_CREDENTIALS_DURATION_SECONDS = 3600;
+    public static final Integer STS_CREDENTIALS_DURATION_SECONDS = 3600;
```
Comment (andresrc) [inline src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: No need to change the type here, as it will be boxed when needed. Besides, I'm not sure if it would be backwards compatible. Better to revert this change.

Comment (nathandines) [reply src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: Ah, I changed type because I was getting boxing and unboxing compilation errors with the new `setStsTokenDuration` logic

Comment (nathandines) [reply src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: I’ll see what I can do to revert it when I get back to a computer

Comment (andresrc) [reply src/main/java/com/cloudbees/jenkins/plugins/awscredentials/AWSCredentialsImpl.java]: Worst case: 
```
this.stsTokenDuration = stsTokenDuration == null || stsTokenDuration.intValue() == STS_CREDENTIALS_DURATION_SECONDS ? null : stsTokenDuration;
```

**Comment by nathandines** [comment]
@andresrc I kept hitting boxing and unboxing errors, so I've reverted to accessing the variable through `DescriptorImpl`. I think this is cleaner than having `.intValue()` all over the place, or boxing the constant in the `set` and `get` methods
